loadbalancer/fadein_test.go

```diff
@@ -178,16 +198,109 @@ func TestFadeIn(t *testing.T) {
 	testFadeIn(t, "random, 8", newRandom, 0, 0, 0, 0, 0, 0)
 	testFadeIn(t, "random, 9", newRandom, fadeInDuration/2, fadeInDuration/3, fadeInDuration/4)
 
-	testFadeIn(t, "consistent-hash, 0", newConsistentHash, old, old)
-	testFadeIn(t, "consistent-hash, 1", newConsistentHash, 0, old)
-	testFadeIn(t, "consistent-hash, 2", newConsistentHash, 0, 0)
-	testFadeIn(t, "consistent-hash, 3", newConsistentHash, old, 0)
-	testFadeIn(t, "consistent-hash, 4", newConsistentHash, old, old, old, 0)
-	testFadeIn(t, "consistent-hash, 5", newConsistentHash, old, old, old, 0, 0, 0)
-	testFadeIn(t, "consistent-hash, 6", newConsistentHash, old, 0, 0, 0)
-	testFadeIn(t, "consistent-hash, 7", newConsistentHash, old, 0, 0, 0, 0, 0, 0)
-	testFadeIn(t, "consistent-hash, 8", newConsistentHash, 0, 0, 0, 0, 0, 0)
-	testFadeIn(t, "consistent-hash, 9", newConsistentHash, fadeInDuration/2, fadeInDuration/3, fadeInDuration/4)
+	testFadeIn(t, "consistent-hash, 0", newConsistentHashForTest, old, old)
+	testFadeIn(t, "consistent-hash, 1", newConsistentHashForTest, 0, old)
+	testFadeIn(t, "consistent-hash, 2", newConsistentHashForTest, 0, 0)
+	testFadeIn(t, "consistent-hash, 3", newConsistentHashForTest, old, 0)
+	testFadeIn(t, "consistent-hash, 4", newConsistentHashForTest, old, old, old, 0)
+	testFadeIn(t, "consistent-hash, 5", newConsistentHashForTest, old, old, old, 0, 0, 0)
+	testFadeIn(t, "consistent-hash, 6", newConsistentHashForTest, old, 0, 0, 0)
+	testFadeIn(t, "consistent-hash, 7", newConsistentHashForTest, old, 0, 0, 0, 0, 0, 0)
+	testFadeIn(t, "consistent-hash, 8", newConsistentHashForTest, 0, 0, 0, 0, 0, 0)
+	testFadeIn(t, "consistent-hash, 9", newConsistentHashForTest, fadeInDuration/2, fadeInDuration/3, fadeInDuration/4)
+}
+
+func testFadeInLoadBetweenOldEps(
+	t *testing.T,
+	name string,
+	algorithm func([]string) routing.LBAlgorithm,
+	nOld int, nNew int,
+) {
+	t.Run(name, func(t *testing.T) {
+		old := fadeInDurationHuge
+		new := time.Duration(0)
+		endpointAges := []time.Duration{}
+		for i := 0; i < nOld; i++ {
+			endpointAges = append(endpointAges, old)
+		}
+		for i := 0; i < nNew; i++ {
+			endpointAges = append(endpointAges, new)
+		}
+
+		ctx, eps := initializeEndpoints(endpointAges, fadeInDurationHuge)
+
+		a := algorithm(eps)
+		rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+		nReqs := map[string]int{}
+
+		t.Log("test start", time.Now())
+		// Emulate the load balancer loop, sending requests to it with random hash keys
+		// over and over again till fadeIn period is over.
+		for i := 0; i < numberOfReqs; i++ {
+			ctx.Params[ConsistentHashKey] = strconv.Itoa(rnd.Intn(100000))
+			ep := a.Apply(ctx)
+			nReqs[ep.Host]++
+		}
+
+		oldEpsExpectedLoad := numberOfReqs / nOld
+		for idx, ep := range eps {
+			if endpointAges[idx] == old {
+				assert.InEpsilon(t, oldEpsExpectedLoad, nReqs[ep], 0.2)
+			}
+			if endpointAges[idx] == new {
+				assert.InDelta(t, 0, nReqs[ep], float64(oldEpsExpectedLoad)*0.05)
```

Comment:
Shouldn't this just be close to zero? I.e. why do we use oldEpsExpectedLoad for _new_ endpoints?

Reply:
I need somehow to define *absolute* error (because relative doesn't work for zero). Decided to create one more constant `acceptableErrorNearZero`.